src/libsyntax/ext/tt/macro_check.rs

```diff
@@ -0,0 +1,342 @@
+                state = 3;
+            }
+            // macro_rules ! . $meta
+            (2, &TokenTree::MetaVar(..)) => {
+                state = 3;
+                check_occurrences(sess, tt, macros, binders, ops, valid);
+            }
+            // macro_rules ! ident . {delim}
+            // macro_rules ! $meta . {delim}
+            // macro ident . {delim}
+            // macro $meta . {delim}
+            (3, &TokenTree::Delimited(_, ref del)) | (5, &TokenTree::Delimited(_, ref del))
+                if del.delim == DelimToken::Brace =>
+            {
+                state = 0;
+                let rest = check_nested_macro(sess, &del.tts, &nested_macros, valid);
+                // If we did not check the whole macro definition, then check the rest as if outside
+                // a macro definition.
+                check_nested_occurrences(sess, &del.tts[rest..], macros, binders, ops, valid);
+            }
+            // macro . ident
+            (4, &TokenTree::Token(Token { kind: TokenKind::Ident(..), .. })) => {
+                state = 5;
+            }
+            // macro . $meta
+            (4, &TokenTree::MetaVar(..)) => {
+                state = 5;
+                check_occurrences(sess, tt, macros, binders, ops, valid);
+            }
+            // macro ident . (delim)
+            // macro $meta . (delim)
+            (5, &TokenTree::Delimited(_, ref del)) if del.delim == DelimToken::Paren => {
+                state = 6;
+                nested_binders = Binders::default();
+                check_binders(sess, tt, &nested_macros, &mut nested_binders, &[], valid);
+            }
+            // macro ident (delim) . {delim}
+            // macro $meta (delim) . {delim}
+            (6, &TokenTree::Delimited(_, ref del)) if del.delim == DelimToken::Brace => {
+                state = 0;
+                check_occurrences(sess, tt, &nested_macros, &nested_binders, &[], valid);
+            }
+            (_, ref tt) => {
+                state = 0;
+                check_occurrences(sess, tt, macros, binders, ops, valid);
+            }
+        }
+    }
+}
+
+fn check_nested_macro(
+    sess: &ParseSess,
+    tts: &[TokenTree],
+    macros: &[MacroState<'_>],
+    valid: &mut bool,
+) -> usize {
+    let n = tts.len();
+    let mut i = 0;
+    loop {
+        if i + 2 >= n
+            || !is_delimited(&tts[i])
+            || !is_token(&tts[i + 1], &TokenKind::FatArrow)
+            || !is_delimited(&tts[i + 2])
+        {
+            break;
+        }
+        let lhs = &tts[i];
+        let rhs = &tts[i + 2];
+        let mut binders = Binders::default();
+        check_binders(sess, lhs, macros, &mut binders, &[], valid);
+        check_occurrences(sess, rhs, macros, &binders, &[], valid);
+        i += 3;
+        if i == n || !is_token(&tts[i], &TokenKind::Semi) {
+            break;
+        }
+        i += 1;
+    }
+    i
+}
+
+fn is_delimited(tt: &TokenTree) -> bool {
```

Comment:
Perhaps just make this an inherent method on `TokenTree`? Same for `is_token`. This can be left for a followup PR if you want (just leave a FIXME or something)...

Reply:
Done.